File: EnglishLearningApp/Sources/Views/Components.swift
import SwiftUI

// MARK: - Card container

struct Card: View {
    var accent: Color
    private let content: AnyView

    init<C: View>(accent: Color = Theme.primary, @ViewBuilder content: () -> C) {
        self.accent = accent
        self.content = AnyView(content())
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            content
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Theme.cardBackground)
        .cornerRadius(16)
        .overlay(
            RoundedRectangle(cornerRadius: 16)
                .stroke(accent.opacity(0.35), lineWidth: 1.5)
        )
    }
}

// MARK: - Section header

struct SectionHeader: View {
    let title: String
    var icon: String = "book.fill"

    var body: some View {
        HStack(spacing: 8) {
            Image(systemName: icon)
                .foregroundColor(Theme.primary)
            Text(title)
                .font(.headline)
                .fontWeight(.bold)
        }
        .padding(.horizontal)
        .padding(.top, 8)
    }
}

// MARK: - Multiple-choice question card (shared by Exercise & Reading)

struct MCQCard: View {
    let question: MCQuestion
    let accent: Color
    var onResult: ((Bool) -> Void)? = nil

    @State private var selectedIndex: Int?
    @State private var didCheck = false

    private var isCorrect: Bool { selectedIndex == question.correctIndex }

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            Text(question.prompt)
                .font(.headline)
                .fixedSize(horizontal: false, vertical: true)

            ForEach(question.options.indices, id: \.self) { index in
                optionRow(index: index)
            }

            if didCheck {
                resultView
            } else {
                Button(action: check) {
                    Text("Check answer")
                        .fontWeight(.semibold)
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 12)
                        .background(Theme.primary)
                        .cornerRadius(12)
                }
                .disabled(selectedIndex == nil)
                .opacity(selectedIndex == nil ? 0.5 : 1)
            }
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Theme.cardBackground)
        .cornerRadius(16)
        .overlay(
            RoundedRectangle(cornerRadius: 16)
                .stroke(accent.opacity(0.35), lineWidth: 1.5)
        )
    }

    private func optionRow(index: Int) -> some View {
        let letter = String(UnicodeScalar(65 + index)!) // A, B, C, D
        let isSelected = selectedIndex == index
        let isCorrectOption = index == question.correctIndex

        var background = Theme.cardBackground
        var border = Color.gray.opacity(0.35)
        var letterColor = Color.secondary

        if didCheck {
            if isCorrectOption {
                background = Theme.success.opacity(0.15)
                border = Theme.success
                letterColor = Theme.success
            } else if isSelected {
                background = Theme.danger.opacity(0.15)
                border = Theme.danger
                letterColor = Theme.danger
            }
        } else if isSelected {
            background = Theme.primary.opacity(0.15)
            border = Theme.primary
            letterColor = Theme.primary
        }

        return Button(action: { select(index) }) {
            HStack(spacing: 12) {
                Text(letter)
                    .font(.headline)
                    .foregroundColor(letterColor)
                    .frame(width: 28, height: 28)
                    .background(letterColor.opacity(0.15))
                    .cornerRadius(14)

                Text(question.options[index])
                    .foregroundColor(.primary)
                    .fixedSize(horizontal: false, vertical: true)

                Spacer()

                if didCheck && isCorrectOption {
                    Image(systemName: "checkmark.circle.fill")
                        .foregroundColor(Theme.success)
                } else if didCheck && isSelected {
                    Image(systemName: "xmark.circle.fill")
                        .foregroundColor(Theme.danger)
                }
            }
            .padding(12)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(background)
            .cornerRadius(12)
            .overlay(
                RoundedRectangle(cornerRadius: 12)
                    .stroke(border, lineWidth: 1.5)
            )
        }
        .buttonStyle(PlainButtonStyle())
    }

    private var resultView: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack(spacing: 8) {
                Image(systemName: isCorrect ? "checkmark.circle.fill" : "xmark.circle.fill")
                Text(isCorrect ? "Correct!" : "Not quite")
                    .fontWeight(.bold)
            }
            .foregroundColor(isCorrect ? Theme.success : Theme.danger)

            Text(question.explanation)
                .font(.subheadline)
                .foregroundColor(.secondary)
                .fixedSize(horizontal: false, vertical: true)

            Button(action: reset) {
                Text("Try again")
                    .font(.footnote)
                    .fontWeight(.semibold)
                    .foregroundColor(Theme.primary)
            }
        }
    }

    private func select(_ index: Int) {
        guard !didCheck else { return }
        selectedIndex = index
    }

    private func check() {
        guard selectedIndex != nil else { return }
        didCheck = true
        onResult?(isCorrect)
    }

    private func reset() {
        selectedIndex = nil
        didCheck = false
    }
}

File: android/app/src/main/java/com/example/englishlearning/ui/components/Components.kt
package com.example.englishlearning.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Cancel
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.example.englishlearning.data.model.MCQuestion
import com.example.englishlearning.ui.theme.PrimaryBlue
import com.example.englishlearning.ui.theme.SuccessGreen
import com.example.englishlearning.ui.theme.DangerRed

/** Rounded, bordered card container — mirrors the iOS `Card` view. */
@Composable
fun AppCard(
    accent: Color,
    modifier: Modifier = Modifier,
    content: @Composable ColumnScope.() -> Unit
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(16.dp))
            .background(MaterialTheme.colorScheme.surface)
            .border(1.5.dp, accent.copy(alpha = 0.35f), RoundedCornerShape(16.dp))
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
        content = content
    )
}

@Composable
fun SectionHeader(title: String, icon: ImageVector = Icons.Filled.CheckCircle) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp)
    ) {
        Icon(icon, contentDescription = null, tint = PrimaryBlue)
        Text(title, style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.Bold)
    }
}

@Composable
fun ProgressBar(value: Double, modifier: Modifier = Modifier) {
    val fraction = value.coerceIn(0.0, 1.0).toFloat()
    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(8.dp)
            .clip(RoundedCornerShape(4.dp))
            .background(Color.Gray.copy(alpha = 0.2f))
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth(fraction)
                .fillMaxHeight()
                .background(PrimaryBlue, RoundedCornerShape(4.dp))
        )
    }
}

@Composable
fun LoadingIndicator(modifier: Modifier = Modifier) {
    Box(modifier = modifier, contentAlignment = Alignment.Center) {
        CircularProgressIndicator()
    }
}

/**
 * Multiple-choice question card with instant marking and an explanation.
 * Mirrors the iOS `MCQCard` shared by the Exercise and Reading sections.
 */
@Composable
fun McqCard(
    question: MCQuestion,
    accent: Color,
    modifier: Modifier = Modifier,
    onResult: ((Boolean) -> Unit)? = null
) {
    var selectedIndex by remember(question.id) { mutableStateOf<Int?>(null) }
    var didCheck by remember(question.id) { mutableStateOf(false) }
    val isCorrect = selectedIndex == question.correctIndex

    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(16.dp))
            .background(MaterialTheme.colorScheme.surface)
            .border(1.5.dp, accent.copy(alpha = 0.35f), RoundedCornerShape(16.dp))
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp)
    ) {
        Text(question.prompt, style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)

        question.options.forEachIndexed { index, option ->
            OptionRow(
                index = index,
                option = option,
                selectedIndex = selectedIndex,
                didCheck = didCheck,
                correctIndex = question.correctIndex,
                onSelect = { if (!didCheck) selectedIndex = index }
            )
        }

        if (didCheck) {
            ResultView(
                isCorrect = isCorrect,
                explanation = question.explanation,
                onReset = {
                    selectedIndex = null
                    didCheck = false
                }
            )
        } else {
            Button(
                onClick = {
                    if (selectedIndex != null) {
                        didCheck = true
                        onResult?.invoke(isCorrect)
                    }
                },
                enabled = selectedIndex != null,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text("Check answer", fontWeight = FontWeight.SemiBold)
            }
        }
    }
}

@Composable
private fun OptionRow(
    index: Int,
    option: String,
    selectedIndex: Int?,
    didCheck: Boolean,
    correctIndex: Int,
    onSelect: () -> Unit
) {
    val letter = ('A'.code + index).toChar()
    val isSelected = selectedIndex == index
    val isCorrectOption = index == correctIndex

    var background = MaterialTheme.colorScheme.surface
    var border = Color.Gray.copy(alpha = 0.35f)
    var letterColor = Color.Gray

    when {
        didCheck && isCorrectOption -> {
            background = SuccessGreen.copy(alpha = 0.15f)
            border = SuccessGreen
            letterColor = SuccessGreen
        }
        didCheck && isSelected -> {
            background = DangerRed.copy(alpha = 0.15f)
            border = DangerRed
            letterColor = DangerRed
        }
        isSelected -> {
            background = PrimaryBlue.copy(alpha = 0.15f)
            border = PrimaryBlue
            letterColor = PrimaryBlue
        }
    }

    Surface(
        onClick = onSelect,
        shape = RoundedCornerShape(12.dp),
        color = background,
        border = androidx.compose.foundation.BorderStroke(1.5.dp, border),
        modifier = Modifier.fillMaxWidth()
    ) {
        Row(
            modifier = Modifier.padding(12.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            Box(
                modifier = Modifier
                    .size(28.dp)
                    .clip(CircleShape)
                    .background(letterColor.copy(alpha = 0.15f)),
                contentAlignment = Alignment.Center
            ) {
                Text(letter.toString(), fontWeight = FontWeight.Bold, color = letterColor)
            }
            Text(option, modifier = Modifier.weight(1f))
            when {
                didCheck && isCorrectOption -> Icon(
                    Icons.Filled.CheckCircle,
                    contentDescription = null,
                    tint = SuccessGreen
                )
                didCheck && isSelected -> Icon(
                    Icons.Filled.Cancel,
                    contentDescription = null,
                    tint = DangerRed
                )
            }
        }
    }
}

@Composable
private fun ResultView(isCorrect: Boolean, explanation: String, onReset: () -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Icon(
                if (isCorrect) Icons.Filled.CheckCircle else Icons.Filled.Cancel,
                contentDescription = null,
                tint = if (isCorrect) SuccessGreen else DangerRed
            )
            Text(
                if (isCorrect) "Correct!" else "Not quite",
                fontWeight = FontWeight.Bold,
                color = if (isCorrect) SuccessGreen else DangerRed
            )
        }
        Text(
            explanation,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
        TextButton(onClick = onReset) {
            Text("Try again", color = PrimaryBlue, fontWeight = FontWeight.SemiBold)
        }
    }
}
